server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java

```diff
@@ -304,12 +307,58 @@ private synchronized void updateSkipUnavailable(String clusterAlias, Boolean ski
+    public synchronized void updateRemoteClusterCredentials(Supplier<Settings> settingsSupplier, ActionListener<Void> listener) {
+        updateRemoteClusterCredentials(settingsSupplier.get(), listener);
     }
 
-    public RemoteClusterCredentialsManager getRemoteClusterCredentialsManager() {
-        return remoteClusterCredentialsManager;
+    public synchronized void updateRemoteClusterCredentials(Settings settings, ActionListener<Void> listener) {
+        final UpdateRemoteClusterCredentialsResult result = remoteClusterCredentialsManager.updateClusterCredentials(settings);
+        // We only need to rebuild connections when a credential was newly added or removed for a cluster alias, not if the credential
+        // value was updated. Therefore, only consider added or removed aliases
+        final int totalConnectionsToRebuild = result.addedClusterAliases().size() + result.removedClusterAliases().size();
+        if (totalConnectionsToRebuild == 0) {
+            logger.debug("no connection rebuilding required after credentials update");
+            listener.onResponse(null);
+            return;
+        }
+        logger.info("rebuilding [{}] connections after credentials update", totalConnectionsToRebuild);
+        final GroupedActionListener<Void> groupedListener = new GroupedActionListener<>(
+            totalConnectionsToRebuild,
+            listener.map(responses -> {
+                logger.info("rebuild complete for [{}] connections after credentials update", responses.size());
+                return null;
+            })
+        );
+        for (var clusterAlias : result.addedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+        for (var clusterAlias : result.removedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+    }
+
+    private void maybeRebuildConnectionOnCredentialsChange(
+        String clusterAlias,
+        Settings settings,
+        GroupedActionListener<Void> groupedListener
+    ) {
+        if (remoteClusters.containsKey(clusterAlias)) {
+            updateRemoteCluster(clusterAlias, settings, true, ActionListener.wrap(status -> {
+                logger.info("remote cluster connection [{}] updated after credentials change: [{}]", clusterAlias, status);
+                groupedListener.onResponse(null);
+            }, e -> {
+                logger.warn(() -> "failed to update remote cluster connection [" + clusterAlias + "] after credentials change", e);
+                // We don't want to return an error to the upstream listener since a connection rebuild failure does not count as
+                // a failed credential reload; same as for remote cluster settings updates, we "handle" the failure by logging a WARN
+                // message
+                groupedListener.onResponse(null);
```

Comment:
Are you sure you want to suppress exceptions here? AFAICT if we propagate the exception then it'll be logged and returned to the caller of `POST /_nodes/reload_secure_settings` which seems desirable.

Also I'd suggest using a `RefCountingListener` since `GroupedActionListener` allocates an unnecessary array to hold all the results. If you really do want to suppress exceptions then use a `RefCountingRunnable` instead.

Also also `ActionListener#wrap` seems unnecesary since the `onResponse` shouldn't be throwing anything. Just make a `new ActionListener<>`.

Reply:
Thanks for the pointers! I'm tweaking this to use a RefCounting-based approach for sure.

I've gone back and forth on whether to surface the exception or not: 

My main reason for not surfacing it is that a connection rebuild failure is not exactly a reload secure settings failure -- when a connection rebuild fails, the credentials reload itself still succeeded (i.e., we are using the new credential going forward) so it could be confusing to get a ["reload failed" log](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java#L136) and error in the `reload-secure-settings` response. This is also aligned with how cluster setting updates work, i.e., updating remote cluster settings does not fail even if the remote connection set up fails.   

For RCS 2.0 in particular, there is a fairly common case where we expect an exception - if a user is migrating from the old security model to the new (or vice versa), they can first update (and reload the credential), then change the target port for the remote connection via a cluster settings update call. In this case, they would get an exception on the reload (since upon rebuilding the connection, the cluster will attempt to connect on the wrong port). Returning an exception here could create a red herring for end-users.  

It's certainly possible to build an exception message that makes the failure cause clear and isn't too verbose (the exceptions have rather verbose stack traces, and there may be several connections with exceptions)  -- pragmatically speaking, I'm not sure it's worth the effort (and slight dissonance with this not _really_ being a reload secure settings failure) since any user that needs to trouble-shoot remote cluster setup will have to rely on logs today in any case. It's a bit of a toss up for me here TBH. Given this context, do you have a strong preference?

Reply:
No strong preference, no. I'd suggest adding a code comment about why we're choosing to suppress failures in this area (largely duplicating your previous message) to help future readers.